src/flp/types.rs

```diff
@@ -314,38 +314,73 @@ impl<F: FftFriendlyFieldElement> Type for Average<F> {
 
 /// The histogram type. Each measurement is an integer in `[0, length)` and the aggregate is a
 /// histogram counting the number of occurrences of each measurement.
-#[derive(Clone, PartialEq, Eq)]
-pub struct Histogram<F: FftFriendlyFieldElement> {
+#[derive(PartialEq, Eq)]
+pub struct Histogram<F, S> {
     length: usize,
-    range_checker: Vec<F>,
+    chunk_len: usize,
+    gadget_calls: usize,
+    phantom: PhantomData<(F, S)>,
 }
 
-impl<F: FftFriendlyFieldElement> Debug for Histogram<F> {
+impl<F: FftFriendlyFieldElement, S> Debug for Histogram<F, S> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("Histogram")
             .field("length", &self.length)
+            .field("chunk_len", &self.chunk_len)
             .finish()
     }
 }
 
-impl<F: FftFriendlyFieldElement> Histogram<F> {
+impl<F: FftFriendlyFieldElement, S: ParallelSumGadget<F, Mul<F>>> Histogram<F, S> {
     /// Return a new [`Histogram`] type with the given number of buckets.
-    pub fn new(length: usize) -> Result<Self, FlpError> {
+    pub fn new(length: usize, chunk_len: usize) -> Result<Self, FlpError> {
         if length >= u32::MAX as usize {
             return Err(FlpError::Encode(
                 "invalid length: number of buckets exceeds maximum permitted".to_string(),
             ));
         }
+        if length == 0 {
+            return Err(FlpError::InvalidParameter(
+                "length cannot be zero".to_string(),
+            ));
+        }
+        if chunk_len == 0 {
```

Comment:
nit: `chunk_length` might be a better name because it lines up better with `length`.

Here and for the other parallel sum types.